torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,138 @@ namespace rpc {
+  rpcRetryMap_[newTime] = std::make_shared<RpcRetryInfo>(
+      to, std::move(message), fm, future, 1, maxRetries);
+  rpcRetryMapCV_.notify_one();
+
+  return future;
+}
+
+void RpcAgent::retryExpiredRpcs() {
+  while (rpcAgentRunning_.load()) {
+    std::unique_lock<std::mutex> lock(rpcRetryMutex_);
 
-RpcAgent::~RpcAgent() = default;
+    // We must continue sleeping as long as the RPC Agent is running and when
+    // either the Retry Map is empty, or when the Retry Map's earliest expiring
+    // RPC is set to be retried in the future.
+    if (rpcRetryMap_.empty()) {
+      rpcRetryMapCV_.wait(lock, [&, this] {
+        return !(rpcAgentRunning_.load() && rpcRetryMap_.empty());
+      });
+    }
+
+    steady_clock_time_point earliestTimeout;
+
+    if (!rpcRetryMap_.empty()) {
+      earliestTimeout = rpcRetryMap_.begin()->first;
+      rpcRetryMapCV_.wait_until(lock, earliestTimeout, [&, this] {
+        return !(rpcAgentRunning_.load());
+      });
+    }
+
+    if (!rpcAgentRunning_.load()) {
+      return;
+    }
+    auto& earliestRpc = rpcRetryMap_[earliestTimeout];
+
+    // If the future for this RPC's last try was not completed or resulted in
+    // an error, we will re-try sending the RPC. We also attach a callback that
+    // marks the original future (which was returned to the user) as completed
+    // is the retry succeeds. Lastly we compute a new time in the future (based
+    // on exponential backoff) at which this RPC would be retried next, and we
+    // update our map accordingly.
+    if (shouldRetryRpc(earliestRpc->retryFuture_)) {
+      auto fm = send(earliestRpc->to_, std::move(earliestRpc->message_));
+      steady_clock_time_point newTime =
+          computeNewRpcRetryTime(earliestRpc->retryCount_);
+
+      fm->addCallback([this, earliestRpc, newTime](
+                          const rpc::Message& message,
+                          const c10::optional<utils::FutureError>& futErr) {
+        if (futErr) {
+          LOG(INFO) << "Send try "
+                    << std::to_string(earliestRpc->retryCount_ + 1)
+                    << " failed\n";
+        } else if (!earliestRpc->originalFuture_->completed()) {
+          // This try succeeded and the original future hasn't been marked
+          // completed yet, so we can mark it as complete.
+          earliestRpc->originalFuture_->markCompleted(message);
+          {
+            std::lock_guard<std::mutex> retryMapLock(rpcRetryMutex_);
+            rpcRetryMap_.erase(newTime);
+          }
+        }
+      });
+
+      earliestRpc->retryCount_ += 1;
+
+      if (earliestRpc->retryCount_ < earliestRpc->maxRetries_) {
+        // Update the RpcRetryInfo struct in-place and then move the
+        // earliestRpc pointer to a new key corresponding to the next
+        // retry time.
+        earliestRpc->retryFuture_ = fm;
+        rpcRetryMap_[newTime] = std::move(earliestRpc);
+        rpcRetryMap_.erase(earliestTimeout);
```

Comment:
Unless I'm missing something it seems like if we exceed maxRetries we don't mark the original future as failed and we don't remove the future from rpcRetryMap_ either.